sdk/tests/core/test_az_mqtt5_telemetry_producer_hfsm.c

```diff
@@ -0,0 +1,430 @@
+                                   .reason_code = 135 }),
+      AZ_OK);
+
+  assert_int_equal(ref_pub_rsp, 1);
+  assert_int_equal(ref_telemetry_prod_err_rsp, 1);
+}
+
+static void test_az_mqtt5_telemetry_producer_send_begin_bad_arg_failure(void** state)
+{
+  (void)state;
+  reset_test_counters();
+
+  az_mqtt5_telemetry_producer_send_req_event_data test_telemetry_data
+      = { .content_type = AZ_SPAN_EMPTY,
+          .qos = AZ_MQTT5_QOS_AT_LEAST_ONCE,
+          .telemetry_payload = AZ_SPAN_FROM_STR(TEST_PAYLOAD),
+          .telemetry_name = AZ_SPAN_FROM_STR(TEST_TELEMETRY_NAME) };
+
+  assert_int_equal(
+      az_mqtt5_telemetry_producer_send_begin(&test_telemetry_producer, &test_telemetry_data),
+      AZ_ERROR_ARG);
+
+  assert_int_equal(ref_pub_req, 0);
+}
+
+static void test_az_mqtt5_telemetry_producer_send_begin_qos_0_success(void** state)
+{
+  (void)state;
+  reset_test_counters();
+
+  az_mqtt5_telemetry_producer_send_req_event_data test_telemetry_data
+      = { .content_type = AZ_SPAN_FROM_STR(TEST_CONTENT_TYPE),
+          .qos = AZ_MQTT5_QOS_AT_MOST_ONCE,
+          .telemetry_payload = AZ_SPAN_FROM_STR(TEST_PAYLOAD),
+          .telemetry_name = AZ_SPAN_FROM_STR(TEST_TELEMETRY_NAME) };
+
+  assert_int_equal(
+      az_mqtt5_telemetry_producer_send_begin(&test_telemetry_producer, &test_telemetry_data),
+      AZ_OK);
+
+  assert_int_equal(ref_pub_req, 1);
+}
+
+static void test_az_mqtt5_telemetry_producer_double_send_qos_0_success(void** state)
+{
+  (void)state;
+  reset_test_counters();
+
+  az_mqtt5_telemetry_producer_send_req_event_data test_telemetry_data
+      = { .content_type = AZ_SPAN_FROM_STR(TEST_CONTENT_TYPE),
+          .qos = AZ_MQTT5_QOS_AT_MOST_ONCE,
+          .telemetry_payload = AZ_SPAN_FROM_STR(TEST_PAYLOAD),
+          .telemetry_name = AZ_SPAN_FROM_STR(TEST_TELEMETRY_NAME) };
+
+  assert_int_equal(
+      az_mqtt5_telemetry_producer_send_begin(&test_telemetry_producer, &test_telemetry_data),
+      AZ_OK);
+
+  assert_int_equal(ref_pub_req, 1);
+
+  assert_int_equal(
+      az_mqtt5_telemetry_producer_send_begin(&test_telemetry_producer, &test_telemetry_data),
+      AZ_OK);
+
+  // 2 pubs should be sent out
+  assert_int_equal(ref_pub_req, 2);
+}
+
+static void test_az_mqtt5_telemetry_producer_send_begin_qos_0_bad_arg_failure(void** state)
+{
+  (void)state;
+  reset_test_counters();
+
+  az_mqtt5_telemetry_producer_send_req_event_data test_telemetry_data
+      = { .content_type = AZ_SPAN_EMPTY,
+          .qos = AZ_MQTT5_QOS_AT_MOST_ONCE,
+          .telemetry_payload = AZ_SPAN_FROM_STR(TEST_PAYLOAD),
+          .telemetry_name = AZ_SPAN_FROM_STR(TEST_TELEMETRY_NAME) };
+
+  assert_int_equal(
+      az_mqtt5_telemetry_producer_send_begin(&test_telemetry_producer, &test_telemetry_data),
+      AZ_ERROR_ARG);
+
+  assert_int_equal(ref_pub_req, 0);
+}
+
+static void test_az_mqtt5_telemetry_producer_send_begin_qos_1_timeout(void** state)
```

Comment:
Should we have a test to see if we can send a qos 0 message while waiting for a qos 1 message to return?